src/controller/CHIPDeviceController.cpp

```diff
@@ -231,6 +231,42 @@ CHIP_ERROR DeviceController::Init(NodeId localDeviceId, ControllerInitParams par
     return err;
 }
 
+CHIP_ERROR DeviceController::GenerateOperationalCertificates(const ByteSpan & CSR, NodeId deviceId, uint8_t * certBuf,
+                                                             uint32_t certBufSize, uint32_t & outCertLen)
```

Comment:
I believe we now have MutableByteSpan for 'modifiable buffer with length'. Let's use it.

Reply:
I tried using `MutableByteSpan` but ran into a few problem.
1. The size of `MutableByteSpan` cannot be changed. So I had to provide an output length along with `MutableByteSpan`.
2. The `MutableByteSpan` cannot be converted seamlessly to a `ByteSpan`. So a lot of manual conversion was still needed to use the returned value as a `ByteSpan`.

The code required too much back and forth conversion from `ScopedMemoryBuffer` -> `MutableByteSpan` -> `ByteSpan`. I think it'll be great if we could stop passing the buffer pointers, length and returned length on function calls. But more work is needed in the Span/ScopedBuffer APIs to make it seamless.

Reply:
https://github.com/project-chip/connectedhomeip/pull/7708 should make MutableByteSpan nicer to use.  As a followup we can then start using it.

To address the `ScopedMemoryBuffer` thing, we could have a `ScopedSizedMemoryBuffer` which takes the size as a template arg, has a no-arg allocator, and auto-converts to a `MutableByteSpan` of the right size or so.

Or we could make `ScopedMemoryBuffer` be that templated-on-size thing, I guess.

Reply:
In this particular instance, certBuf and certBufSize seem to be input only and the outCertLen is the output, so it seems already split out.

This only passes the argument into ConvertX509CertToChipCert which in turn only uses it to `Init` and ASN1Writer. I do not understand the casting complaint - you would need a constructor when calling "Generate" however I do not see other casting/resizing/other conversions right now.

For general usage, casting between mutable and non-mutable as well as resizing sounds indeed useful. For this particular instance though, can't we just use mutable?

Reply:
I see below a few more instances where indeed passing scoped memory buffer to things that could take 'MutableByteSpan' would also be more readable.

Would take it one at a time still: if we can do this in one function, lets do that and leave only others for more code clarity passes. Just because inconvenient in some places should not mean 'do not use anywhere'.